packages/gatsby-plugin-page-creator/src/create-pages-from-changed-nodes.ts

```diff
@@ -0,0 +1,148 @@
+import { createPath } from "gatsby-page-utils"
+import { reverseLookupParams } from "./extract-query"
+import { getCollectionRouteParams } from "./get-collection-route-params"
+import { derivePath } from "./derive-path"
+import type { CreatePagesArgs } from "gatsby"
+import sysPath from "path"
+import { getPluginInstance, IStatePerInstance } from "./tracked-nodes-state"
+import { IOptions } from "./types"
+
+function createPathFromNode({ filePath, node, reporter, slugifyOptions }): {
+  path: string
+  errors: number
+} {
+  const { derivedPath, errors } = derivePath(
+    filePath,
+    node,
+    reporter,
+    slugifyOptions
+  )
+  return { path: createPath(derivedPath), errors }
+}
+
+export function createPagesFromChangedNodes(
+  {
+    actions,
+    reporter,
+  }: Pick<CreatePagesArgs, "actions" | "reporter"> & {
+    pluginInstance: IStatePerInstance
+  },
+  pluginOptions: IOptions
+): void {
+  // Loop over deleted/created nodes and delete nodes and create nodes
+  // we haven't seen before and then reset arrays.
+  const pluginInstance = getPluginInstance(pluginOptions)
+
+  if (pluginInstance.trackedTypes.size === 0) {
+    return
+  }
+
+  for (const {
+    node,
+  } of pluginInstance.changedNodesSinceLastPageCreation.deleted.values()) {
+    if (pluginInstance.trackedTypes.has(node.internal.type)) {
+      const absolutePaths = pluginInstance.trackedTypes.get(node.internal.type)
+      if (absolutePaths) {
+        for (const absolutePath of absolutePaths) {
+          const { path } = createPathFromNode({
+            filePath: sysPath.relative(pluginOptions.path, absolutePath),
+            node,
```

Comment:
this `node` can't be just a raw node as it is right now as it won't handle resolved nodes

we also won't be able to run query for this node, because it doesn't exist anymore

so instead we have to maintain a map of node id to pages created from that node so we could just lookup pages that needs deleting when node by ids is removed

Reply:
done in https://github.com/gatsbyjs/gatsby/pull/36623/commits/051fd2b812a35b0785f9fb3cd7a0161e5207b7e6